src/homework3/task4.py

```diff
@@ -0,0 +1,39 @@
+"""Homework 3 - Task 4"""
+
+from math import factorial
+
+
+def pairs_number(lst: list) -> int:
+    """Пары элементов
+
+    Дан список чисел. Посчитайте, сколько в нем пар элементов, равных друг другу.
+    Считается, что любые два элемента, равные друг другу образуют одну пару, которую необходимо
+    посчитать.
+    Входные данные: строка из чисел, разделенная пробелами.
+    Выходные данные: количество пар.
+    Важно: 1 1 1 - это 3 пары, 1 1 1 1 - это 6 пар
+    """
+    ret_val = 0
+    elements_counts = {}
+    for element in lst:
+        elements_counts[element] = elements_counts.get(element, 0) + 1
+    for count in elements_counts.values():
+        if count > 1:
+            ret_val += combinations_without_repetitions(count, 2)
+    return ret_val
+
+
+def combinations_without_repetitions(n: int, k: int) -> int:
+    """Число сочетаний без повторений из n по к
+
+    Генерирует исключение ValueError, если n < k.
+    """
+    if n < k:
+        raise ValueError('n < k')
+    return int(factorial(n) / (factorial(n - k) * factorial(k)))
```

Comment:
O(n!) complexity. has O(1) solution

Reply:
So far I am at a loss with the solution of O(1).
How to make the O(1) if we need to bypass the input list at least once? => O(n)

Reply:
it can be O(1) instead of O(n!) for counting - just reworked combinations. we surely have O(n) for read